### PR TITLE
feat: Implement Facade for Update Personal Details

### LINKED ISSUE #1145: 'Update Personal Details' Facade Functions
See "Facade Tickets" in GH-1580

### PR DESCRIPTION
**Pull request by znikola**
closes #1145

### CODE REVIEW

File: projects/core/src/process/process.module.ts [hunk 1]
```diff
@@ -1,5 +1,4 @@
 import { NgModule } from '@angular/core';
-
```
Comment (plabadie) [inline projects/core/src/process/process.module.ts]: Can changed in this file be reverted?

File: projects/core/src/process/store/process-store.module.ts [hunk 2]
```diff
@@ -1,12 +1,8 @@
 import { NgModule } from '@angular/core';
-
 import { StoreModule } from '@ngrx/store';
-
 import { StateModule } from '../../state/state.module';
-
 import { PROCESS_FEATURE } from './process-state';
-
-import { reducerToken, reducerProvider } from './reducers/index';
+import { reducerProvider, reducerToken } from './reducers/index';
```
Comment (plabadie) [inline projects/core/src/process/store/process-store.module.ts]: Can this change be reverted?

File: projects/core/src/user/facade/user.service.spec.ts [hunk 3]
```diff
@@ -400,6 +404,60 @@ describe('UserService', () => {
...
+        .subscribe(loading => (result = loading))
+        .unsubscribe();
+
+      expect(result).toEqual(true);
+    });
+
+    it('should return the error flag', () => {
+      spyOn(service, 'getUpdatePersonalDetailsResultError').and.returnValue(
+        of(true)
+      );
```
Comment (plabadie) [inline projects/core/src/user/facade/user.service.spec.ts]: You can't mock the function being tested.

File: projects/core/src/user/facade/user.service.spec.ts [hunk 4]
```diff
@@ -400,6 +404,60 @@ describe('UserService', () => {
...
+      service.updatePersonalDetails(username, userDetails);
+      expect(store.dispatch).toHaveBeenCalledWith(
+        new fromStore.UpdateUserDetails({ username, userDetails })
+      );
+    });
+
+    it('should return the loading flag', () => {
+      spyOn(service, 'getUpdatePersonalDetailsResultLoading').and.returnValue(
+        of(true)
+      );
```
Comment (plabadie) [inline projects/core/src/user/facade/user.service.spec.ts]: You can't mock the function being tested.

File: projects/core/src/user/store/user-state.ts [hunk 5]
```diff
@@ -11,6 +11,8 @@ import {
 } from '../../occ/occ-models/index';
 
 export const USER_FEATURE = 'user';
+export const USER_UPDATE_PROCESS = 'updateUserDetails';
```
Comment (plabadie) [inline projects/core/src/user/store/user-state.ts]: Could we match the the actions nomenclature and call the process something like `PROCESS_UPDATE_USER_DETAILS`.  I think it would improve the readability of the code that uses the constant.
Lets discuss.

File: projects/core/src/user/occ/user.service.spec.ts [hunk 6]
```diff
@@ -76,6 +76,24 @@ describe('OccUserService', () => {
...
 
+  describe('update user details', () => {
+    it('should update user details for the given username', () => {
+      const userUpdates: User = {
+        title: 'mr',
+      };
+      service.updateUserDetails(username, userUpdates).subscribe(_ => _);
+
+      const mockReq = httpMock.expectOne(req => {
+        return req.method === 'PATCH' && req.url === endpoint + `/${username}`;
```
Comment (plabadie) [inline projects/core/src/user/occ/user.service.spec.ts]: Let's also validate that the new user details are in the request body, or rather that we get the expected request body for a given set of user details.

Having the 3 fields we expose in the form would be nice too.

File: projects/core/src/user/store/effects/user-details.effect.ts [hunk 7]
```diff
@@ -29,6 +29,31 @@ export class UserDetailsEffects {
...
   );
 
+  @Effect()
+  updateUserDetails$: Observable<
+    | fromUserDetailsAction.UpdateUserDetailsSuccess
+    | fromUserDetailsAction.UpdateUserDetailsFail
+  > = this.actions$.pipe(
+    ofType(fromUserDetailsAction.UPDATE_USER_DETAILS),
+    map((action: fromUserDetailsAction.UpdateUserDetails) => action.payload),
+    // TODO:#1145 - test how the update behaves with `concatMap` and with `switchMap`
```
Comment (plabadie) [inline projects/core/src/user/store/effects/user-details.effect.ts]: Resolve or remove the TODO

File: projects/core/src/user/store/reducers/index.ts [hunk 8]
```diff
@@ -1,35 +1,39 @@
...
 import * as fromUserDetailsReducer from './user-details.reducer';
 import * as fromUserOrdersReducer from './user-orders.reducer';
-import * as fromResetPasswordReducer from './reset-password.reducer';
 
 export function getReducers(): ActionReducerMap<UserState> {
   return {
-    account: fromUserDetailsReducer.reducer,
+    account: combineReducers({
+      details: fromUserDetailsReducer.reducer,
+    }),
```
Comment (plabadie) [inline projects/core/src/user/store/reducers/index.ts]: Do we still need to update the user state considering we store the process state elsewhere?

File: projects/core/src/user/store/reducers/user-details.reducer.ts [hunk 9]
```diff
@@ -1,24 +1,28 @@
...
 export function reducer(
   state = initialState,
   action: fromUserDetailsAction.UserDetailsAction
-): UserDetailsState {
+): User {
   switch (action.type) {
     case fromUserDetailsAction.LOAD_USER_DETAILS_SUCCESS: {
-      const details: User = action.payload;
+      return action.payload;
+    }
```
Comment (plabadie) [inline projects/core/src/user/store/reducers/user-details.reducer.ts]: What is the purpose of this change in LOAD_USER_DETAILS_SUCCESS reducer?

File: projects/core/src/user/occ/user.service.spec.ts (hunk 6 above)
Comment (znikola) [reply projects/core/src/user/occ/user.service.spec.ts]: Are you referring to this? https://github.com/SAP/cloud-commerce-spartacus-storefront/pull/1873/files/3a714095a99cb20833e08ab3cdb8a272c3494d19#diff-5697e5d8eaa2e2e3dd18de4e952de5f8R92

File: projects/core/src/user/store/user-state.ts (hunk 5 above)
Comment (znikola) [reply projects/core/src/user/store/user-state.ts]: Done

File: projects/core/src/user/store/reducers/index.ts (hunk 8 above)
Comment (znikola) [reply projects/core/src/user/store/reducers/index.ts]: Yes, we're merging the state locally.

File: projects/core/src/user/facade/user.service.spec.ts (hunk 4 above)
Comment (znikola) [reply projects/core/src/user/facade/user.service.spec.ts]: oops

File: projects/core/src/user/store/reducers/user-details.reducer.ts (hunk 9 above)
Comment (znikola) [reply projects/core/src/user/store/reducers/user-details.reducer.ts]: My guess would be when we load the user details for the first time, we store it under `user.details` in the state.

File: projects/core/src/user/occ/user.service.spec.ts (hunk 6 above)
Comment (plabadie) [reply projects/core/src/user/occ/user.service.spec.ts]: Yes, that's what I'm referring to : )
